Core/include/Acts/TrackFitting/Chi2Fitter.hpp

```diff
@@ -0,0 +1,767 @@
+      if (sourcelink_it != inputMeasurements->end()) {
+        ACTS_VERBOSE("chi2 |    processSurface: Measurement surface "
+                     << surface->geometryId() << " detected.");
+
+        // add a full TrackState entry multi trajectory
+        // (this allocates storage for all components, we will set them later)
+        result.lastTrackIndex = result.fittedStates->addTrackState(
+            ~(TrackStatePropMask::Smoothed | TrackStatePropMask::Filtered),
+            result.lastTrackIndex);
+
+        // now get track state proxy back
+        auto trackStateProxy =
+            result.fittedStates->getTrackState(result.lastTrackIndex);
+
+        trackStateProxy.setReferenceSurface(surface->getSharedPtr());
+
+        // assign the source link to the track state
+        trackStateProxy.setUncalibrated(sourcelink_it->second);
+
+        // Fill the track state
+        trackStateProxy.predicted() = std::move(boundParams.parameters());
+        if (boundParams.covariance().has_value()) {
+          trackStateProxy.predictedCovariance() =
+              std::move(*boundParams.covariance());
+        }
+        trackStateProxy.jacobian() = std::move(jacobian);
+        trackStateProxy.pathLength() = std::move(pathLength);
+
+        extensions.calibrator(state.geoContext, trackStateProxy);
+
+        visit_measurement(trackStateProxy.calibratedSize(), [&](auto N) {
+          constexpr size_t kMeasurementSize = decltype(N)::value;
+
+          // simple projection matrix H_is, composed of 1 and 0, 2x6 or 1x6
+          const ActsMatrix<kMeasurementSize, eBoundSize> proj =
+              trackStateProxy.projector()
+                  .template topLeftCorner<kMeasurementSize, eBoundSize>();
+
+          const auto Hi =
+              (proj * result.jacobianFromStart).eval();  // 2x6 or 1x6
+          const auto localMeasurements =
+              trackStateProxy
+                  .template calibrated<kMeasurementSize>();  // 2x1 or 1x1
+
+          const auto& covariance =
+              trackStateProxy.template calibratedCovariance<
+                  kMeasurementSize>();  // 2x2 or 1x1. Should
+                                        // be diagonal.
+          const auto covInv = covariance.inverse();
+
+          auto residuals =
+              localMeasurements - proj * trackStateProxy.predicted();
+
+          // TODO: use detail::calculateResiduals? Theta/Phi?
+          const auto deriv1 = (-2 * Hi.transpose() * covInv * residuals).eval();
+          const auto deriv2 = (2 * Hi.transpose() * covInv * Hi).eval();
+          result.collectorDeriv1Sum += deriv1;
+          result.collectorDeriv2Sum += deriv2;
+
+          for (int i = 0; i < localMeasurements.rows(); ++i) {
+            result.collectorMeasurements.push_back(localMeasurements(i));
+            result.collectorResiduals.push_back(residuals(i));
+            result.collectorCovariance.push_back(covariance(i, i));
```

Comment:
I missed this: are all residuals, measurement and covariances collected in vector(s)  which is then used for the minimisation ?